galsim/phase_psf.py

```diff
@@ -0,0 +1,1017 @@
+                self.rng == other.rng)
+
+    def __ne__(self, other):
+        return not self == other
+
+    def advance(self):
+        """Advance phase screen realization by self.time_step."""
+        # For ARAtmosphericScreen, since we use Fourier methods to update the screen each step
+        # anyway, it's easy to simultaneously move the screen with the wind.
+        if self.noise_frac < 1.e-12:
+            # Frozen flow (but slower than FrozenAtmosphericScreen, since using Fourier methods to
+            # move the screen in the wind).
+            self._phaseFT = self.alpha*self._phaseFT
+        else:
+            # Boiling, do a fractional random phase update.
+            self._phaseFT = self.alpha*self._phaseFT + self._noiseFT()*self.noise_frac
+        self.screen = np.fft.ifft2(self._phaseFT).real
+
+    def path_difference(self, nx, scale, theta_x=0.0*galsim.degrees, theta_y=0.0*galsim.degrees):
+        """ Compute effective pathlength differences due to phase screen.
+
+        @param nx       Size of output array
+        @param scale    Scale of output array pixels in meters
+        @param theta_x  Field angle corresponding to center of output array.
+        @param theta_y  Ditto.
+        @returns   Array of pathlength differences in nanometers.  Multiply by 2pi/wavelength to get
+                   array of phase differences.
+        """
+        img = galsim.Image(np.ascontiguousarray(self.screen), scale=self.screen_scale)
+        ii = galsim.InterpolatedImage(img, calculate_stepk=False, calculate_maxk=False,
+                                      normalization='sb')
+        ii = ii.shift(1000*self.altitude * theta_x.tan(), 1000*self.altitude * theta_y.tan())
+        return ii.drawImage(nx=nx, ny=nx, scale=scale, method='sb', dtype=np.float64).array
+
+    def reset(self):
+        """Reset phase screen back to time=0."""
+        self.rng = copy.deepcopy(self.orig_rng)
+        fx, fy = self._freq()
+        self._init_screen(fx, fy)
+
+
+class PhaseScreenList(object):
+    """ List of phase screens that can be turned into a PSF.  Screens can be either atmospheric
+    layers or optical phase screens.  Generally, one would assemble a PhaseScreenList object using
+    the function `Atmosphere`.  Layers can be added, removed, appended, etc. just like items can be
+    manipulated in a python list.  For example:
+
+    # Create an atmosphere with three layers, two of which are frozen and one not frozen.
+    > screens = galsim.PhaseScreenList([galsim.FrozenAtmosphericScreen(...),
+                                        galsim.FrozenAtmosphericScreen(...),
+                                        galsim.ARAtmosphericScreen(...)])
+    # Add another layer
+    > screens.append(galsim.FrozenAtmosphericScreen(...))
+    # Remove the second layer
+    > del screens[1]
+    # Switch the first and second layer.  Silly, but works...
+    > screens[0], screens[1] = screens[1], screens[0]
```

Comment:
Another minor point.  Our style for these has been to tab these in 4 spaces and use `>>>` for the code bits, not just `>`.  The idea is to match the appearance of an interactive python session.